definition_validator/variance.cc

```diff
@@ -0,0 +1,184 @@
+
+#include "common/typecase.h"
+#include "core/Symbols.h"
+#include "core/core.h"
+#include "core/errors/resolver.h"
+
+#include "variance.h"
+
+using namespace std;
+
+namespace sorbet::definition_validator::variance {
+
+class VarianceValidator {
+private:
+    const core::Loc loc;
+
+    VarianceValidator(const core::Loc loc) : loc(loc) {}
+
+    core::Variance invertVariance(const core::Variance variance) {
+        switch (variance) {
+            case core::Variance::CoVariant:
+                return core::Variance::ContraVariant;
+            case core::Variance::Invariant:
+                return core::Variance::Invariant;
+            case core::Variance::ContraVariant:
+                return core::Variance::CoVariant;
+        }
+    }
+
+    string showVariance(const core::Variance variance) {
+        switch (variance) {
+            case core::Variance::CoVariant:
+                return ":out";
+            case core::Variance::Invariant:
+                return ":invariant";
+            case core::Variance::ContraVariant:
+                return ":in";
+        }
+    }
+
+    bool compatibleVariance(const core::Variance left, const core::Variance right) {
+        switch (left) {
+            case core::Variance::CoVariant:
+                return right != core::Variance::ContraVariant;
+            case core::Variance::Invariant:
+                return right == core::Variance::Invariant;
+            case core::Variance::ContraVariant:
+                return right != core::Variance::CoVariant;
+        }
+    }
+
+    void validate(const core::Context ctx, const core::Variance variance, const core::TypePtr type) {
+        typecase(
+            type.get(), [&](core::ClassType *klass) {},
+
+            [&](core::LiteralType *lit) {},
+
+            [&](core::SelfType *self) {},
+
+            [&](core::SelfTypeParam *sp) {},
+
+            [&](core::TypeVar *tvar) {},
+
+            [&](core::OrType *any) {
+                validate(ctx, variance, any->left);
+                validate(ctx, variance, any->right);
+            },
+
+            [&](core::AndType *all) {
+                validate(ctx, variance, all->left);
+                validate(ctx, variance, all->right);
+            },
+
+            [&](core::AliasType *alias) {
+                auto aliasData = alias->symbol.data(ctx);
```

Comment:
ooc: why does this even ever see alias types?
Do we have a test for this?

Reply:
if it does, the right thing to do is dealias + validate.

Reply:
If you put `module_function` in a module with type variables, you'll see this case.